Multids

rminnich opened the pull request:
You can now easily run more than one command with a single decpu command, e.g.
./decpu -n 2 -key ~/.ssh/cpu_rsa -sp 17020 . date
or
./decpu -n 2 -key ~/.ssh/cpu_rsa -sp 17020 dnssd://?arch=amd64  date
Sat Aug  4 10:29:11 AM UTC 2012
Thu Jun 21 01:28:22 PM UTC 2012

./decpu -n 2 -key ~/.ssh/cpu_rsa -sp 17020 dnssd://?arch=amd64 hostname
atomicpi
api2

ericvh commented on ds/ds.go:
```diff
@@ -275,6 +276,56 @@ func Lookup(query dsQuery) (string, string, error) {
+// LookupChan returns a chan of *LookupResult and a cancel function.
+// The cancel function can be called when the desired number of results has
+// been received; or for any other reason in fact.
+func LookupChan(query dsQuery) (chan *LookupResult, context.CancelFunc) {
```
Take a look at my resubmitted sort patch -- the original intended function was to sort the list of responses before returning.  While this was previously used to just pick the top one (which usually was the one with the lowest load because that was the default sort parameter) -- you probably still want to pick the top n from that list before getting a list of all the possible ones.  Logic could be here in Lookup or in decpu, I'm flexible on that, but I think we need to integrate sorting into whatever solution we end up with.

rminnich replied:
for now, let's start with just matching for now, sorting can come later. 

The go way to do this would be to
func LookupChan(query dsQuery, opts ...func([]*LookupResult) error) {
..
get the host list as []*LookupResult
for o := range opts {
if err := o(l); err != nil {
return err
}
}

And those opts would bring in sorting or whatever else we need. That way, default is not sorted, but sorting can be added.

ericvh commented on cmds/decpu/decpu.go:
```diff
@@ -207,13 +209,23 @@ func main() {
+			for e := range c {
+				amt++
+				if amt > *numCPUs {
+					return
```
okay, so if we were to move sorting to the client, instead of instantiating the cpus on every loop, we wait till we hit n, sort the list and then instantiate.  I suppose the downside is with the code the way it is we dispatch as soon as the remote tells us and otherwise we are waiting for our timeout before dispatching.  We could conditionalize holding off on dispatching ONLY if we detect sort -- but we'd need parse to return a bool telling us whether or not there is a sort involved (by default there is though).

ericvh reviewed (commented):
Need to work out how to re-integrate sort which complicates this logic slightly. Otherwise looks good.

abrender commented on ds/ds.go:
```diff
@@ -275,6 +276,56 @@ func Lookup(query dsQuery) (string, string, error) {
+// LookupChan returns a chan of *LookupResult and a cancel function.
+// The cancel function can be called when the desired number of results has
+// been received; or for any other reason in fact.
+func LookupChan(query dsQuery) (chan *LookupResult, context.CancelFunc) {
```
Just pointing out "Rethinking Classical Concurrency Patterns" by bcmills - (video: https://www.youtube.com/watch?v=5zXAHh5tJqQ slides:   https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view).    Exposing channels as part of the public API of this package is risky and ambiguous.  Whose job is it to close the channel - the caller or the callee **(explicitely: can you please document the expected behavior (who will close the channel) in the docstring? thanks!)?**

rminnich replied:
done. Thanks.

The reason for the chan is that one might overlap getting lookup results and initiating the cpu connections. This is a common pattern in HPC. 

In this case, I doubt the lookup will take that long, so I'm okay with the changes.

ericvh commented on ds/ds.go:
```diff
@@ -357,6 +358,67 @@ func Lookup(query dsQuery) (string, string, error) {
+			break
+		}
+	}
+	return ret, err
```
Wouldn't we just need the dsSort before return like in standard Lookup() to get the sorted responses?

rminnich replied:
The problem I have with that is -- sorted by what key? I'm a bit leery of embedding a sort in here without knowing what the user wants to sort on?

rminnich replied:
anyway ... I view the sort as another PR if you don't mind, makes bisect better

rminnich replied:
btw the idiomatic way to do this ... you'd change it as follows:
type Opt func([]*LookupResult) ([]*LookupResult, error)

func LookupN(query dsQuery, n int, opts ...Opt) ([]*LookupResult, error) {

right before the return:
for _, o := range opts {
   ret, err = o(ret)
   if err != nil {
    return ret, err
}

}

That lets you refine in any way you choose.
I'd rather do that than wire in a sort anyway.

ericvh replied:
So ... here's the thing.  dsSort was already in the code as part of Lookup, you are essentially removing the sort functionality by switching to LookupN as the default, but you don't even change the README to remove the sort options nor do you remove the sort tests in TESTDOCKERCPU -- so its actually you that is breaking the bisect.

rminnich replied:
LookupN was not in the package before, and the sorts were deadcode.

I'll close this PR, go ahead and do what you think best, but as it is it's hard to get "more than 1" from a lookup.

ericvh commented:
the keys are specified in the URI and there are defaults based on tenants
and load average (this is the way it worked before lookupN) but I’ll submit
in a separate PR.

On Thu, Dec 29, 2022 at 11:08 AM ron minnich ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In ds/ds.go
> <https://github.com/u-root/cpu/pull/177#discussion_r1059064038>:
>
> > +		close(respCh)
> +	}()
> +
> +	var ret []*LookupResult
> +	var err error
> +	for l := range respCh {
> +		if l.Error != nil {
> +			err = l.Error
> +			break
> +		}
> +		ret = append(ret, l)
> +		if len(ret) >= n {
> +			break
> +		}
> +	}
> +	return ret, err
>
> btw the idiomatic way to do this ... you'd change it as follows:
> type Opt func([]*LookupResult) ([]*LookupResult, error)
>
> func LookupN(query dsQuery, n int, opts ...Opt) ([]*LookupResult, error) {
>
> right before the return:
> for _, o := range opts {
> ret, err = o(ret)
> if err != nil {
> return ret, err
> }
>
> }
>
> That lets you refine in any way you choose.
> I'd rather do that than wire in a sort anyway.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/u-root/cpu/pull/177#discussion_r1059064038>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAQ7WDJFLP4CDYGKGYXCW3WPXASNANCNFSM6AAAAAASTF2VEE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

ericvh commented:
It wasn't deadcode, it was called in regular Lookup here: https://github.com/u-root/cpu/blob/fa5cb43f3cd637f2060f6273bd42114e078e11a6/ds/ds.go#L351

But you don't need to close the PR, just change where the sort happens to the client.  Hold up and I'll send you a PR for your PR.

ericvh commented:
see [feature/dsLookupN](https://github.com/u-root/cpu/tree/feature/dsLookupN) for how to fix

rminnich committed:
cmds/decpu: support more than one remote host

This required some changes to ds.go, since, as written, it would always return
the same value for the host, no matter how many, as it was using a pointer to a
LookupResult and overwriting it.

The chan in LookupResult was really not needed, and neither was the goroutine.

Now, with these changes:
rminnich@a300:~/go/src/github.com/u-root/cpu/cmds/decpu$ ./decpu  -n 3 -key ~/.ssh/cpu_rsa -sp 17020 dnssd://?arch=amd64 hostname
atomicpi
api3
rminnich@a300:~/go/src/github.com/u-root/cpu/cmds/decpu$ ./decpu  -n 1 -key ~/.ssh/cpu_rsa -sp 17020 dnssd://?arch=amd64 hostname
api3
rminnich@a300:~/go/src/github.com/u-root/cpu/cmds/decpu$ ./decpu  -n 2 -key ~/.ssh/cpu_rsa -sp 17020 dnssd://?arch=amd64 hostname
api3
atomicpi

Note there's a price to be paid in performance for the dnssd lookup:

rminnich@a300:~/go/src/github.com/u-root/cpu/cmds/decpu$ time cpu atomicpi date
Tue Jul 24 04:57:24 AM UTC 2012

real	0m0.373s
user	0m0.120s
sys	0m0.097s

rminnich@a300:~/go/src/github.com/u-root/cpu/cmds/decpu$ time ./decpu  -n 1 -key ~/.ssh/cpu_rsa -sp 17020 dnssd://?arch=amd64 hostname
atomicpi

real	0m1.726s
user	0m1.905s
sys	0m0.258s

Note, you can run both daemons on the same machine: the decpud was started as follows, to a
host with u-root:
cpu api3 cp decpud /bbin
PATH=$PATH:/bbin cpu api3 decpud -sp 17020 -pk ~/.ssh/cpu_rsa.pub -d

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

rminnich committed:
Merge branch 'main' into multids

rminnich committed:
Merge branch 'main' into multids

rminnich commented:
by the power vested in me as the creator of this project, I am approving this PR.
